File: KisaragiTech.Dape/User/Model/LocalRegisteredUser.cs
using KisaragiTech.Dape.User.Interface;
using KisaragiTech.Dape.User.Service;

namespace KisaragiTech.Dape.User.Model;

/// <summary>
/// 本登録したローカルユーザー。必要な権限を持っていれば各種操作が行えるようになる。
/// </summary>
/// <param name="identifier"></param>
/// <param name="preferredHandle"></param>
public sealed class LocalRegisteredUser(UserID identifier, string preferredHandle, HashedPassword hashedPassword) : ILocalUser
{
    public UserID GetIdentifier()
    {
        return identifier;
    }

    public string GetPreferredHandle()
    {
        return preferredHandle;
    }

    public bool VerifyPassword(string raw)
    {
        return hashedPassword.Equals(PasswordHasher.CreateHashedPassword(raw));
    }
}

File: src/main/scala/io/github/ksrgtech/dape/user/model/LocalRegisteredUser.scala
package io.github.ksrgtech.dape.user.model

import cats.effect.kernel.Sync

/** A fully registered local user who can perform various operations if they have the necessary permissions.
  */
final case class LocalRegisteredUser(
    identifier: UserId,
    preferredHandle: String,
    hashedPassword: HashedPassword
) extends LocalUser {
  override def getIdentifier: UserId = identifier

  def verifyPassword[F[_]: Sync](raw: String): F[Boolean] =
    hashedPassword.verify[F](raw)
}
